keps/sig-storage/1495-volume-populators/README.md

```diff
@@ -0,0 +1,492 @@
+as a `TypedLocalObjectReference` to give flexibility in the future about what objects could be
+data sources for new volumes.
+
+Since then, we have allowed only two things to be the source of a
+new volume -- existing PVCs (indicating a user's intent to clone the volume) and snapshots
+(indicating the user's intent to restore a snapshot). Implementation of these two data sources
+relies on a CSI plugin to do the actual work, and no generic implementation exists for cloning
+volumes or restoring snapshots.
+
+Since the original design of the `DataSource` API field, we have been aware of the desire to
+populate volumes with data from other sources, both CSI-specific and generic (compatible with
+any CSI plugin). For new CSI-specific data sources, the path forward is clear, but for other
+sources of data, which I call "Volume Populators" we don't have a mechanism. The main
+problem is that current API validation logic uses a white list of object types, driven by
+feature gates for each object type. This approach won't scale to generic populators, which
+will by their nature be too numerous and varied.
+
+This proposal recommends that we relax validation (in core Kubernetes) on the `DataSource`
+field to allow arbitrary object types to be data sources, and rely on new controllers to
+perform the validation on those data sources and provide feedback to users. We will introduce
+a new CRD to register valid datasource kinds, and individual volume populators will
+create a CR for each kind of data source that it understands.
+
+## Motivation
+
+### Goals
+
+- Enable users to create pre-populated volumes in a manner consistent with current practice
+- Enable developers to innovate with new an interesting data sources
+- Avoid changing existing workflows and breaking existing tools with new behavior
+- Ensure that users continue to get accurate error messages and feedback about volume
+  creation progress, whether using existing methods of volume creation or specifying a
+  data source associated with a volume populator.
+- Maintain existing pattern of allowing things to happen out of order. As long as there's
+  a chance things will succeed in the future, just provide a status update and retry
+  again later.
+
+### Non-Goals
+
+- This proposal DOES NOT recommend any specific approach for data populators. The specific
+  design of how a data populator should work will be handled in a separate KEP.
+
+## Proposal
+
+Validation for the `DataSource` field will be relaxed so that any API object (except
+core API objects other than PVCs) may be specified. For objects that are not valid data
+source, rather than relying the the API server to reject them or ignore them (the prior
+behavior was to ignore them by clearing the field) allow them and provide feedback to
+API users with events.
+
+A new controller will be introduced called `data-source-validator` which will watch PVCs
```

Comment:
Why a controller and events, rather than a synchronous admission control?  I'd rather know my thing won't work via a hard-fail rather than a success + wait + WTF?  + oh look there's an event.